### ISSUE TITLE
[eslint-plugin-react-hooks] Add option to require functions from core hooks in dependencies

### ISSUE THREAD

**Issue by decafdennis** [facebook/react#15175]
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

The `react-hooks/exhaustive-deps` will currently accept either (and fix neither) of the following:

```javascript
// Scenario A: Function returned by core hook IS NOT specified as a dependency.
const [value, setValue] = useState(initialValue);
const toggle = useCallback(() => setValue(v => !v), []);
```

```javascript
// Scenario B: Function returned by core hook IS specified as a dependency.
const [value, setValue] = useState(initialValue);
const toggle = useCallback(() => setValue(v => !v), [setValue]);
```

**What is the expected behavior?**

If we add the following to our `.eslintrc.js`

```
// Introduces a `requireCoreFunctions` config option.
'react-hooks/exhaustive-deps': ['error', { requireCoreFunctions: true }]
```
then the rule should fail in scenario A, and fixing should result in scenario B.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

n/a

**Comment by gaearon** [comment]
What makes this useful?

**Comment by decafdennis** [comment]
TLDR; it would eliminate a gray area for those who want to be more strict and consistent with their code standards.

With regards to including setState/dispatch as hook inputs/dependencies, you've been saying something to the effect of "you don't have to, but it doesn't hurt." (Correct me if I'm wrong!) That makes sense, but leaves the developer with a decision to make and an opportunity to bikeshed. 

We would prefer to put this in our code standards, and have the linter enforce it:

> If you do specify a `deps` parameter, specify all dependencies, including functions returned from core hooks.

**Comment by decafdennis** [comment]
PS. Would be happy to submit a PR if it would have a chance of being considered. 😄

**Comment by micha149** [comment]
Is `setValue` actually **not** required in the dependency array? I always assumed it is required…

**Comment by decafdennis** [comment]
> I always assumed it is required…

Technically it’s not, it’s optional.

Case in point. :)
